tools/evict-shared-cache/evict-shared-cache.cpp

```diff
@@ -0,0 +1,175 @@
+     << std::endl;
+  // clang-format on
+}
+
+std::unique_ptr<EvictionPolicy> make_policy(const char* argv0, const char* policy) {
+  if (strcmp(policy, "nil") == 0) {
+    return std::make_unique<NilEvictionPolicy>();
+  }
+
+  std::cerr << "Unknown policy: " << policy << std::endl;
+  print_help(argv0);
+  exit(EXIT_FAILURE);
+}
+
+bool stdstreams_still_open() {
+  int status = fcntl(STDIN_FILENO, F_GETFD);
+
+  if (status == -1 && errno == EBADF) {
+    return false;
+  }
+
+  status = fcntl(STDOUT_FILENO, F_GETFD);
+
+  if (status == -1 && errno == EBADF) {
+    return false;
+  }
+
+  return true;
+}
+
+struct CommandParser {
+  std::string command_buff = "";
+
+  std::vector<std::string> read_commands() {
+    std::vector<std::string> commands = {};
+    while (stdstreams_still_open()) {
+      uint8_t buffer[4096] = {};
+
+      ssize_t count = read(STDIN_FILENO, static_cast<void*>(buffer), 4096);
+
+      // Nothing new to process, yield control
+      if (count == 0) {
+        return commands;
+      }
+
+      // An error occured during read
+      if (count < 0) {
+        std::cerr << "Failed to read from stdin: " << strerror(errno) << std::endl;
+        exit(EXIT_FAILURE);
+      }
+
+      // do the stuff
+      uint8_t* iter = buffer;
+      uint8_t* buffer_end = buffer + count;
+      while (iter < buffer_end) {
+        auto end = std::find(iter, buffer_end, 0);
+        command_buff.append(iter, end);
+        if (end != buffer_end) {
+          commands.emplace_back(std::move(command_buff));
+          command_buff = "";
+        }
+        iter = end + 1;
+      }
+
+      // last read consumed the buffer, yield control
+      if (count < 4096) {
+        return commands;
+      }
+
+      // yield if 100 commands have been buffered up without yielding
+      if (commands.size() > 100) {
+        return commands;
+      }
+    }
+    // stdin or stdout were closed. Exit cleanly
+    exit(EXIT_SUCCESS);
+  }
+};
+
+int main(int argc, char** argv) {
+  // clang-format off
+  struct option options[] {
+    {0, "cache", GOPT_ARGUMENT_REQUIRED},
+    {0, "policy", GOPT_ARGUMENT_REQUIRED},
+    {'h', "help", GOPT_ARGUMENT_FORBIDDEN},
+    {0, 0, GOPT_LAST}
+  };
+  // clang-format on
+
+  argc = gopt(argv, options);
+  gopt_errors(argv[0], options);
+
+  bool help = arg(options, "help")->count;
+  const char* cache = arg(options, "cache")->argument;
+  const char* policy_name = arg(options, "policy")->argument;
+
+  if (help) {
+    print_help(argv[0]);
+    exit(EXIT_SUCCESS);
+  }
+
+  if (!cache) {
+    std::cerr << "Cache directory not specified" << std::endl;
+    print_help(argv[0]);
+    exit(EXIT_FAILURE);
+  }
+
+  if (!policy_name) {
+    std::cerr << "Eviction policy not specified" << std::endl;
+    print_help(argv[0]);
+    exit(EXIT_FAILURE);
+  }
+
+  std::unique_ptr<EvictionPolicy> policy = make_policy(argv[0], policy_name);
+  policy->init();
+
+  CommandParser cmd_parser;
+  while (true) {
```

Comment:
this should be while stdin/stdout are open

Reply:
That is checked inside of read commands atm. Didn't  want to double check

Reply:
Left a comment above. I think we should actually use poll instead of checking to see if the file is open (a file being closed is an IO event that Poll will notify you of, if `read` return `EBADF` then the handle is closed). We should also return something like maybe wcl::optional<std::vector<std::string>> so that the empty optional case indicates that stdin was closed.